### PR TITLE
[9] whatsapp chatbot

### LINKED ISSUE #9: Exploring WhatsApp integration with ChatGPT
As part of our mission-related prototyping work, I'm curious to explore how easy or difficult it is to bring together WhatsApp and large language models. 

To begin with, I plan to:

- Scan the solutions for bridging WhatsApp and ChatGPT API (or other LLMs) and write a short note about the existing solutions, their pricing and pros/cons 
- Implement a quick test using one of the solutions

### PR DESCRIPTION
**Pull request by beingkk**
Closes #9 

A prototype to test out OpenAI integration with WhatsApp

Implements a simple Flask API that can receive text messages from WhatsApp/Twilio and send back a response.

It supports two types of responses:
- 'explain' : Sending a simple explanation of a user-provided concept
- 'activities': Suggesting conversation and activity ideas based on user-provided topic

At the moment, this runs successfully using local machine as the server + ngrok to expose the localhost to outside world

I can make a separate PR for setting this up on heroku, as I will need to figure out how to make it work with poetry and custom packages.

### CODE REVIEW

File: whatsapp_bot.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    for text in texts:
+        client.messages.create(body=text, from_=my_contact, to=receiver_contact)
+        sleep(0.5)
+    return
+
+
+@app.route("/")
+def hello() -> str:
+    """Test the api is working"""
+    return "Hello, World!"
```
Comment (kstathou) [inline whatsapp_bot.py]: Shall we remove this once we're done testing?

File: whatsapp_bot.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        sender_contact=request.form.get("From"),
+        receiver_contact=request.form.get("To"),
+    )
+    resp = MessagingResponse()
+    resp.message(reply)
+    return str(resp)
+
+
+if __name__ == "__main__":
+    app.run(debug=True)
```
Comment (kstathou) [inline whatsapp_bot.py]: We should remove the `debug=True` flag before merging this PR

**Review by kstathou — APPROVED** [review]
LGTM!

**Comment by beingkk** [comment]
Hey @kstathou, it would be great to get your advice on a couple things:

- How to best organise the code: I've tried to simplify things by using your utils (`ActivityGenerator`) and having one 'whatsapp_bot.py' file for the Flask API app. However, for some reason I find it a bit unsettling to have prompt complication in the same file as the API functions - let me know what you think?

- Conflicting pre-commit hooks: When I try to commit, `Flake8` complains about whitespace:
```
whatsapp_bot.py:136:25: E203 whitespace before ':'
```
However, when I delete the whitspace, `black` complains about formatting and restores the whitespace. So, seems like they're at odds... What would you suggest to do?
The line causing the issue is the following:
```
texts = [text_body[i : i + 1500] for i in range(0, len(text_body), 1500)]
```

- FYI so far I've tested this with localhost + ngrok. I will also need to figure out how to make poetry and custom packages (ie, `genai`) work with Heroku, but happy to do that as a separate PR. Also, I can then write a readme about how to link it to Twilio, as I'll be writing a blog about it.

File: whatsapp_bot.py (hunk 2 above)
Comment (beingkk) [reply whatsapp_bot.py]: Will do!

**Comment by beingkk** [comment]
Oh, another FYI @kstathou - I've slightly adjusted [one of the messages of the activity generation prompt](https://github.com/nestauk/discovery_generative_ai/pull/10/files#diff-6ae68b6c7346b25495890e7b17ef20d5ed7261a529e970849982ff438e2fd811). 

Sometimes the OpenAI response doesn't neatly separate Conversations and Games and Crafts, but instead alternates between one Conversation and one Games and Crafts idea. The same behaviour happened in the streamlit app as well sometimes.

So, I adjusted this part of the prompt to make the instructions more explicit:
> ```Please propose two conversation ideas and several games and crafts ideas I could do with the children to extend their learning.\n\nTypes of activities:\n- Conversations: Asking them questions about the topic\n- Games and Crafts: Puzzles, games, role play, arts and crafts\n\n###Formatting###\nReturn the proposed activities in the following format and always start with the two conversation ideas and then follow with all games and crafts ideas```

**Pull request merged**